pyat/at/errors/error_def.py

```diff
@@ -0,0 +1,449 @@
+        rr = ring.replace(refpts)
+        for e in rr[refpts]:
+            setattr(e, pname, get_pol(e, pname, pstat, pdyn, index))
+            sanitize(e)
+        return rr
+
+    alldef = kwargs.pop('all', True)
+    for pol in ['A', 'B']:
+        if kwargs.pop('Polynom'+pol, alldef):
+            index = kwargs.pop('Index'+pol, None)
+            ring = set_polerr(ring, 'Polynom'+pol, index)
+    return ring
+    
+
+def enable_errors(ring: Lattice, **kwargs):
+    r"""Enable magnet errors
+
+    Returns a shallow copy of *ring* where magnet errors defined by
+    :py:func:`assign_errors` are introduced in the magnet attributes.
+
+    By default, all assigned errors are enabled, but keyword arguments allow a
+    selection of the enabled errors.
+
+    Args:
+        ring:   Lattice description.
+
+    Keyword Args:
+        all (bool):         Set the default value for all the specific error
+          flags. Default: :py:obj:`True`
+        shiftx (bool):      enable horizontal shift errors. Default: *all*
+        shifty (bool):      enable vertical shift errors. Default: *all*
+        tilt (bool):        enable tilt errors. Default: *all*
+        pitch (bool):       enable pitch errors. Default: *all*
+        yaw (bool):         enable yaw errors. Default: *all*
+        PolynomB (bool):    enable polynomial errors. Default: *all*
+        IndexB (Optional[int]): restrict the polynomial errors to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+        PolynomA (bool):    enable polynomial errors. Default: *all*
+        IndexA (Optional[int]): restrict the skew polynomial error to the
+          specified index. Default: :py:obj:`None` meaning the whole polynom
+
+    Examples:
+        >>> ringerr = enable_errors(ring)
+
+        All errors are enabled in *ringerr*
+
+        >>> ringerr = enable_errors(all=False, ShiftErr=True)
+
+        Only magnet shift errors are active
+
+        >>> ringerr = enable_errors(all=False)
+
+        No magnet errors are enabled. Only monitor errors will be active.
+
+    See also:
+        :py:func:`assign_errors`, :py:func:`get_optics_err`
+    """
+    if not ring.has_errors:
+        ring = _apply_field_errors(ring, **kwargs)
+        ring = _apply_alignment_errors(ring, **kwargs)
+        ring._has_errors = True
+    return ring
+
+
+def has_errors(ring):
+    """Get the error status of a lattice"""
+    return getattr(ring, '_has_errors', False)
+
+
+def get_mean_std_err(ring: Lattice, key, attr, index=0):
+    vals = [np.atleast_1d(getattr(e, attr, 0.0))[index]
+            for e in ring.get_elements(key)]
+    return np.mean(vals), np.std(vals)
+
+
+def _sort_flags(kwargs):
+    errargs = {}
+    for key in _SEL_ARGS:
+        if key in kwargs:
+            errargs[key] = kwargs.pop(key)
+    return kwargs, errargs
+
+
+def find_orbit_err(ring: Lattice, refpts: Refpts = None, **kwargs):
```

Comment:
this function may be used with a lattice without errors or it gives errors? why a new function is needed? I think that it would be a much larger step forward if we could avoid using `*_err`. This startegy was used in the past for private codes. a `find_orbit` sufficiently clever to understand if a lattice has errors or not would be highly user friendly.

Reply:
Hi @simoneliuzzo : the idea is the following:

You may assign errors to a lattice once for ever: they are stored in dedicated attributes and are ignored in all optics computations. So `find_orbit`, `get_optics` and so on are not affected by the "assigned" errors.

The errors are processed only in the `*_err` functions. This avoids confusion: you don't have to care whether assigned errors are present in the lattice or not, they are ignored unless you explicitly call a` *_err` function.

Reply:
I find the choice to use *_err not really user friendly.  I have done like you propose in my code for many years, simply because it would have been a too big change to do it "in the correct way". I expect that find_orbit gives me an orbit with errors, as it does for classic magnet errors T1/T2 R1/R2. Why for BPM errors should it be different? As a user, I would think that find_orbit does not work if I set bpm errors and I see no change in the orbit provided by find_orbit. I would not think (at least immediately) that I am supposed to use a specific functions for the task (that I would then rather call find_orbit_bpm_errors)
I think this point is rather important to be discussed as it changes the present architecture of this pull request. What is the opinion of @swhite2401 , @lcarver, @carmignani, @ThorstenHellert and other users?

Reply:
@simoneliuzzo , in fact my initial intention was to decorate `find_orbit` which would results in the behavior you are seeking for.

There are nevertheless 2 issues with that in the present implementation:
1- it is less efficient since you will always scan through the whole lattice to check for the presence of errors
2- in case you assign errors to an element, calculations will always include then

A possible workaround is to define a lattice attribute `ring.errors_enabled` that one would check before scanning the lattice, it can be done without much effort and thinking about it I do agree with you that such implementation resembling the `enable/disable_6d` functionality would be much nicer for the end user.... now let's see what other users think.

Any arguments against this @lfarv?

Reply:
@simoneliuzzo, @lcarver, @carmignani @ThorstenHellert:

Keep in mind that there are 2 different things: the orbit and the BPM readings. Here there are 2 functions: `find_orbit` gives you the true orbit (affected by T1/T2, R1/R2 and magnet errors, as Simone says, but independent of BPM errors), and `find_orbit_err` which returns the BPM readings.

Maybe the function names could be improved: `find_orbit_view` , `find_optics_view` ?

If there is a strong argument for having a single function, I see 3 possibilities:
- adding another output argument (or field name for `get_optics`) to keep orbit and orbit_view separated. That's the cleanest solution,
- adding a flag `errors_enabled` which will trigger the sequence "enable errors, compute, process BPM errors", as the *_err functions do,
- putting the flag at the lattice level `ring.errors_enabled ` (Simon's proposal), with the same effect. In this case, when should the random errors be generated ? When setting the flag or when starting a computation ?